Improve instructions for new contributors (#1394)
jazzband/pip-tools: CONTRIBUTING.md

to abide by the [Contributor Code of Conduct](https://jazzband.co/about/conduct)
and follow the [guidelines](https://jazzband.co/about/guidelines).

## Getting started

Install tox:

>     pip install tox

Reviewer: Since it's not RST, we should probably use markdown + maybe mark it with a `console` lexer:
```suggestion
```console
$ python -m pip install tox --user
```\
```

Reviewer: Although I'm not sure if this should dive into the details of using tools like pip and tox — maybe it's better to point to the respective docs rather than attempting to maintain copies of their guides?

Author: Markdown seems to understand the format I used though (I agree it's missing the console lexer though)

![image](https://user-images.githubusercontent.com/12172017/116482832-3a2c5500-a886-11eb-9f02-ddf9515664ba.png)

Create a virtual environment, with `pip-tools` in development mode and its test dependencies,
using your current python version:

    tox --devenv .venv -e py

Activate the virtual environment:

>     source ./.venv/bin/activate

Reviewer: I think it's cleaner like this:
```suggestion
    source .venv/bin/activate
```

Author: Agreed

## Project Contribution Guidelines

Here are a few additional or emphasized guidelines to follow when contributing to pip-tools:

- Check with `tox -e checkqa` to see your changes are not breaking the style conventions.

> - Always provide tests for your changes, use `tox -e coverage` to run the tests with coverage reports.

Reviewer: Honestly, running `tox` will run all the test envs with coverage + linters. Not sure if mentioning just `coverage` is needed. Besides, codecov will report back to PRs if it's decreased, I think. Maybe this sort of detail belongs to a development guide rather than contribution instructions. I see such details as non-mandatory things that may help during the development process but should be used if the developer actually needs them. Same with use `tox -e py` to run on test env with the current Python, use `tox -e pyXY` to specify other Python, and so on — we could basically end up documenting how tox works but that would probably be too overwhelming.
Mentioning `tox -av` along the lines might be useful, OTOH.

Author: `tox -e checkqa` was already there when I started contributing to `pip-tools`, so it already looked like a mix of contribution and development guide to me. 
We could eventually regroup these instructions in a Development guide section of this file? I think making another file for this is too much. I personally like a README that tells me exactly what to do to get started quickly. Later I can go to the tool's website to understand it better (we could provide the links too).

Also, when I ran `tox -e coverage` and `tox -e checkqa`, it ran the checks and the tests only for my version of python. Do you suggest that we mention to run `tox` only, to run everything once?

By the way what I did for my first PR:
- Ensured the tests were passing for my python version, checked the code coverage reports about my new tests
- Run the code linters which fixed my files
- See in my PR that the CI succeeded for every python versions

Reviewer 2: Maybe `tox -p all`?

Author: FYI `tox -p all` is really slow on my old macbook air:

```
✔ OK py39-piplatest-coverage in 9 minutes, 36.482 seconds
✔ OK py39-pipprevious-coverage in 10 minutes, 20.35 seconds
✔ OK py39-pip20.3-coverage in 10 minutes, 20.76 seconds
✔ OK py39-pipmain-coverage in 10 minutes, 40.551 seconds
```

So do you guys consider running `tox` with all test environments mandatory?
Also it seems to run only for `py39`, the current version I am using. Does that mean devs should install other python versions on their machine so `tox` can use them as well? If so, this is something we should mention I think? (it's my first time contributing to an open source project and I am not used to how `tox` works)

According to what you say, I suggest to change `Getting started` to `Development Guidelines`. 
Inside, we could suggest command lines to get things done correctly. But it should be kept simple, as @webknjaz said, we don't want to write a full doc about the tools and all possibilities.

Inside the contributing guidelines, we could have:

```
- Always provide tests for your changes.
- Check with `tox` that your changes are not breaking the style conventions.
```

And inside `Development Guidelines`, we could have what we wrote in `Getting started` and add a line or two about testing the branch, probably `$ tox -p all`.

What do you think?

to abide by the [Contributor Code of Conduct](https://jazzband.co/about/conduct)
and follow the [guidelines](https://jazzband.co/about/guidelines).

## Getting started

Install tox:

```console
$ python -m pip install tox --user
```

Create a virtual environment, with `pip-tools` in development mode and its test dependencies,
using your current python version:

```console
$ tox --devenv .venv -e py
```

Activate the virtual environment:

```console

> $ source .venv/bin/activate

Reviewer: One thing I'm concerned about is that it doesn't make it clear that running tests through tox will not use this env, it can be used for development detached from tox and if one wants to run pytest there, they should do so directly.

```

## Project Contribution Guidelines

Here are a few additional or emphasized guidelines to follow when contributing to pip-tools:

> - Check with `tox -e checkqa` to see your changes are not breaking the style conventions.

Reviewer 2: We could drop this line since `pre-commit.ci` checks and autofixes QA issues. What do you think?

Author: Yes that's a nice catch.
Also if we suggest to run `tox -p all` then I think it will still be run?

Reviewer: I think it's a good idea to have `-p` there. The ultimate goal is to let people know how to run the same test matrix that the CI runs locally. And running linters locally can also be useful because people often want to reproduce the offenses in their env to be able to address them. Rerunning the whole CI on every typo fix push is wasteful.

## Project Contribution Guidelines

Here are a few additional or emphasized guidelines to follow when contributing to `pip-tools`:

> - Use `tox` to create a [development environment](https://tox.readthedocs.io/en/latest/example/devenv.html#creating-development-environments-using-the-devenv-option).

Reviewer: Maybe something like this?
```suggestion
- If you need to have a virtualenv outside of `tox`, it is possible to reuse its configuration to provision it as [described in the docs](https://tox.readthedocs.io/en/latest/example/devenv.html#creating-development-environments-using-the-devenv-option).
```

> - Always provide tests for your changes and run `tox -p all` to make sure they are passing the checks.

Reviewer: ```suggestion
- Always provide tests for your changes and run `tox -p all` to make sure they are passing the checks locally.
```

## Project Contribution Guidelines

Here are a few additional or emphasized guidelines to follow when contributing to `pip-tools`:

- To set up your project's environment, have a look at [Python Packaging User Guide](https://packaging.python.org)

>   and [Real Python Tutorials](https://realpython.com)

Reviewer: Could you link to specific articles rather than website homepages?

Author: I couldn't find any good article about `tox` so I provided a link to tox's documentation instead.

> - Install pip-tools in development mode and its test dependencies with `pip install -e '.[testing]'`.

Reviewer: Let's avoid telling people to directly access the extras. We don't have a linter that would enforce a person changing the name of the extra in `setup.cfg` or some setting in `tox.ini` to also make sure to change the guide and keep it in sync at all times. Why did you drop the `--devenv` thing anyway? It would create a venv with deps identical to what tox has.

Author: I now suggest to install tox and use the devenv option specified in tox's documentation. I am only suggesting one command line so we won't have to keep this doc up to date.
